src/Symfony/Component/Runtime/SymfonyRuntime.php

```diff
@@ -0,0 +1,205 @@
+use Symfony\Component\Runtime\Internal\MissingDotenv;
+use Symfony\Component\Runtime\Runner\Symfony\ConsoleApplicationRunner;
+use Symfony\Component\Runtime\Runner\Symfony\HttpKernelRunner;
+use Symfony\Component\Runtime\Runner\Symfony\ResponseRunner;
+
+// Help opcache.preload discover always-needed symbols
+class_exists(ResponseRunner::class);
+class_exists(HttpKernelRunner::class);
+class_exists(MissingDotenv::class, false) || class_exists(Dotenv::class) || class_exists(MissingDotenv::class);
+
+/**
+ * Knows the basic conventions to run Symfony apps.
+ *
+ * Accepts the following options:
+ *  - "debug" to toggle debugging features
+ *  - "env" to define the name of the environment the app runs in
+ *  - "disable_dotenv" to disable looking for .env files
+ *  - "dotenv_path" to define the path of dot-env files - defaults to ".env"
+ *  - "prod_envs" to define the names of the production envs - defaults to ["prod"]
+ *  - "test_envs" to define the names of the test envs - defaults to ["test"]
+ *
+ * When these options are not defined, they will fallback:
+ *  - to reading the "APP_DEBUG" and "APP_ENV" environment variables;
+ *  - to parsing the "--env|-e" and "--no-debug" command line arguments
+ *    if the "symfony/console" component is installed.
+ *
+ * When the "symfony/dotenv" component is installed, .env files are loaded.
+ * When "symfony/error-handler" is installed, it is registred in debug mode.
+ *
+ * On top of the base arguments provided by GenericRuntime,
+ * this runtime can feed the app-callable with arguments of type:
+ *  - Request from "symfony/http-foundation" if the component is installed;
+ *  - Application, Command, InputInterface and/or OutputInterface
+ *    from "symfony/console" if the component is installed.
+ *
+ * This runtime can handle app-callables that return instances of either:
+ *  - HttpKernelInterface,
+ *  - Response,
+ *  - Application,
+ *  - Command,
+ *  - int|string|null as handled by GenericRuntime.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @experimental in 5.3
+ */
+class SymfonyRuntime extends GenericRuntime
+{
+    private $input;
+    private $output;
+    private $console;
+    private $command;
+    private $env;
+
+    /**
+     * @param array {
+     *   debug?: ?bool,
+     *   env?: ?string,
+     *   disable_dotenv?: ?bool,
+     *   project_dir?: ?string,
+     *   prod_envs?: ?string[],
+     *   dotenv_path?: ?string,
+     *   test_envs?: ?string[],
+     * } $options
+     */
+    public function __construct(array $options = [])
+    {
+        $this->env = $options['env'] ?? null;
+        $_SERVER['APP_ENV'] = $options['env'] ?? $_SERVER['APP_ENV'] ?? $_ENV['APP_ENV'] ?? null;
+
+        if (isset($_SERVER['argv']) && null === $this->env && class_exists(ArgvInput::class)) {
+            $this->getInput();
+        }
+
+        if (!($options['disable_dotenv'] ?? false) && isset($options['project_dir']) && !class_exists(MissingDotenv::class, false)) {
```

Comment:
Why dont replace

```php
!class_exists(MissingDotenv::class, false)
```

with 
```php
class_exists(Dotenv::class)
```

Reply:
Because this triggers the autoloader, which is a waste of resources, especially when preloading is used.
Basically, this piece of code is make to work around the fact that preloading/opcache cannot cache the fact that a class does *not* exist.

Reply:
You might be all correct. But Im still confused. 

Isnt this line also triggering autoloading: : https://github.com/symfony/symfony/pull/38465/files#diff-f10a759ae2c9968c992201726dfe5032d3731010bf6832bd82c36459e4eab601R33

```php
class_exists(MissingDotenv::class, false) || class_exists(Dotenv::class) || class_exists(MissingDotenv::class);
```

Reply:
Not when preloading already made `MissingDotenv` xor `Dotenv` available.

Reply:
Oh, that is correct. Since the code in the constructor is not found by the preloader. 
Thank you

Cool. I got it. Btw, maybe add a small comment on the `MissingDotenv` class.